Add share button for lessons

franol-geleta opened the pull request:
- Implement share URL generation for lessons using knowgod.com base URL
- Add share button in lesson UI positioned beside progress bar
- Integrate with existing BaseToolActivity share infrastructure
- Support deferred deep linking for users without the app installed

franol-geleta committed:
Add share button for lessons

- Implement share URL generation for lessons using knowgod.com base URL
- Add share button in lesson UI positioned beside progress bar
- Integrate with existing BaseToolActivity share infrastructure
- Support deferred deep linking for users without the app installed

frett commented on ui/lesson-renderer/src/main/kotlin/org/cru/godtools/tool/lesson/ui/LessonActivity.kt:
```diff
@@ -192,11 +206,31 @@ class LessonActivity :
+                                    )
+                                }
+                            }
+                        }
```
instead of adding it as a composable here, it should be added as an action icon in the `RenderLesson` composable where we actually have the app bar. that can be found in this repo: https://github.com/CruGlobal/kotlin-mpp-godtools-tool-parser

frett reviewed (changes requested):
please move the share icon to the actual app bar for lessons, which is found in our shared kotlin multiplatform project

franol-geleta committed:
Remove overlay share button, keep share URL generation

- Remove overlay share button implementation from LessonActivity
- Keep shareLinkUriLiveData override for share URL generation
- Share button will be added in RenderLesson composable app bar (kotlin-mpp-godtools-tool-parser repo)

franol-geleta committed:
Add share button for lessons in app bar

- Implement share URL generation for lessons using knowgod.com base URL
- Pass share callback to RenderLesson composable to display share button in app bar
- Share button appears beside progress bar when share link is available
- Integrate with existing BaseToolActivity share infrastructure
- Support deferred deep linking for users without the app installed
- Use local shared project build for testing (includeBuild in settings.gradle.kts)

The share button is implemented in the shared Kotlin multiplatform project
(kotlin-mpp-godtools-tool-parser) and integrated here via the onShareClick callback.

frett commented on ui/lesson-renderer/src/main/kotlin/org/cru/godtools/tool/lesson/ui/LessonActivity.kt:
```diff
@@ -193,11 +193,18 @@ class LessonActivity :
+                            state = state,
+                            onShareClick = if (shareLinkUri != null && state is LessonScreen.UiState.Loaded) {
+                                { shareCurrentTool() }
+                            } else null
```
there is already an event callback present in the `LessonScreen.UiState.Loaded` object, you should add an new share link event in there for this callback

franol-geleta committed:
Refactor lesson share to use event-based pattern

- Refactored share functionality to use LessonScreen.UiEvent.ShareLesson event
  instead of callback parameter, following the existing event pattern used by
  CloseLesson event
- Updated eventSink handler to process ShareLesson event and call shareCurrentTool()
  when shareLinkUri is available
- Removed onShareClick callback parameter from RenderLesson composable call
- This change aligns with reviewer feedback to use the existing event callback
  system present in LessonScreen.UiState.Loaded

Note: Corresponding changes are required in kotlin-mpp-godtools-tool-parser:
- Added ShareLesson event to LessonScreen.UiEvent sealed interface
- Updated RenderLesson to use eventSink(ShareLesson) in TopAppBar actions
- Added lesson_accessibility_action_share string resource
- Share button only displays when manifest has both code and locale

frett commented on ui/lesson-renderer/src/main/kotlin/org/cru/godtools/tool/lesson/ui/LessonActivity.kt:
```diff
@@ -321,6 +330,23 @@ class LessonActivity :
+        val locale = locale ?: return null
+        return URI_SHARE_BASE.buildUpon()
+            .appendEncodedPath(locale.toString().lowercase(Locale.ENGLISH))
+            .appendPath("tool")
```
`tool` shouldn't be part of the lesson url, a sample url from our stage site: https://stage.knowgod.com/en/lesson/lessonsharefaith/0

frett commented on settings.gradle.kts:
```diff
@@ -62,6 +62,7 @@ dependencyResolutionManagement {
 rootProject.name = "godtools"
 
 includeBuild("build-logic")
+includeBuild("../kotlin-mpp-godtools-tool-parser")
```
this is good for developing locally, but shouldn't be included in the PR.

We already have the multiplatform project setup to publish dependencies that the android app depends on, once changes to the multiplatform project are approved and merged we can use the updated dependencies.

franol-geleta replied:
noted on this , 


so should i have to push new PR with this thing removed ?

frett replied:
The flow:

- you need to create a PR with the changes made to the multiplatform project
- once that is merged a new build will be published automatically
- you will update the `godtools-shared` version number in `libs.versions.toml` to reference the new version and remove this line
- this PR can be finished being reviewed and merged

frett replied:
you should be able to update the godtoolsShared version in `libs.versions.toml` to `1.3.2-SNAPSHOT` to use the updated code from there

franol-geleta replied:
Done! I've updated the godtoolsShared version in libs.versions.toml to 1.3.2-SNAPSHOT. The change has been committed and pushed to this PR.

franol-geleta committed:
Fix lesson share URL format and remove includeBuild

- Remove 'tool' segment from lesson share URL
- Change URL format from /{locale}/tool/lesson/{tool} to /{locale}/lesson/{tool}/{page}
- Add current page tracking for share links
- Remove includeBuild line for kotlin-mpp-godtools-tool-parser as per PR feedback

franol-geleta committed:
Update LessonActivity to pass showShareAction to UiState

- Pass showShareAction property when creating LessonScreen.UiState.Loaded
- showShareAction is set based on whether shareLinkUri is available
- This works with the multiplatform project changes to separate business logic from UI

franol-geleta committed:
Update godtoolsShared version to 1.3.2-SNAPSHOT

frett reviewed (changes requested):
there are ktlint style checks that are failing and need to be addressed

franol-geleta committed:
Fix all ktlint style violations in lesson-renderer module

- Fix import ordering: Settings, Settings.Companion.*, then URI_SHARE_BASE
- Remove static imports of Intent.ACTION_VIEW from both files
- Replace all ACTION_VIEW references with Intent.ACTION_VIEW
- Format Intent instantiations with proper line breaks to comply with 120 char limit
- Add trailing commas to Intent constructor arguments
- Add blank line between when-conditions in LessonActivity.kt eventSink

Fixes ktlint violations:
- Import ordering: Settings before Settings.Companion.* before URI_SHARE_BASE
- Import ordering: removed static imports
- Line length: formatted Intent calls with line breaks (lines 74-77, 90-93, 106-109, 122-125)
- Spacing: added blank line between when-conditions
- Reference: use Intent.ACTION_VIEW instead of static import
